File: 0061-rotate-list/0061-rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        int size = 0;
        ListNode temp = head;
        while(temp!=null){
            size++;
            temp = temp.next;
        }
        if(size <= 1) return head;
        k = k%size;
        temp = head;
        ListNode temp2 = head;
        
        for(int i = 0; i < size - k-1; i++){
            temp = temp.next;
        }
        if(temp.next == null) return head; // to avoid complete rotation
    
        head = temp.next;
        temp.next = null;
        temp = head;
        while(temp.next!=null){
            temp = temp.next;
        }
        temp.next = temp2;
        System.out.println(size);
        return head;
    }
}

File: 0061-rotate-list/0061-rotate-list.cpp
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(head == NULL) return head;
        int sz = 0;
        ListNode* temp = head;
        while(temp){
            sz++;
            temp = temp->next;
        }
        k = k%sz;
        
        ListNode* fast = head;
        while(k--){
            fast = fast->next;
        }
        ListNode* slow = head;
        while(fast && fast->next){
            fast = fast->next;
            slow = slow->next;
        }
        fast->next = head;
        ListNode* ret = slow->next;
        slow->next = NULL;
        
        return ret;
    }
};